cluster mempool: extend DepGraph functionality

sipa opened the pull request:
Part of cluster mempool: #30289 

This adds:
* `DepGraph::AddDependencies` to add 0 or more dependencies to a single transaction at once (identical to calling `DepGraph::AddDependency` once for each, but more efficient).
* `DepGraph::RemoveTransactions` to remove 0 or more transactions from a depgraph.
* `DepGraph::GetReducedParents` (and `DepGraph::GetReducedChildren`) to get the (reduced) direct parents and children of a transaction in a depgraph.

After which, the `Cluster` type is removed.

This is the result of fleshing out the design for the "intermediate layer" ("TxGraph", no PR yet) between the cluster linearization layer and the mempool layer. My earlier thinking was that TxGraph would store `Cluster` objects (vectors of pairs of `FeeFrac`s and sets of parents), and convert them to `DepGraph` on the fly whenever needed. However, after more consideration, it seems better to have TxGraph store `DepGraph` objects, and manipulate them directly without constantly re-creating them. This requires `DepGraph` to have some additional functionality.

The bulk of the complexity here is the addition of `DepGraph::RemoveTransactions`, which leaves the remaining transactions' positions within the `DepGraph` untouched (we want existing identifiers to remain valid), so this implies that graphs can now have "holes" (positions that are unused, but followed by positions that are used). To enable that, an extension of the fuzz/test serialization format `DepGraphFormatter` is included to deal with such holes.

sipa committed:
clusterlin: make DepGraph::AddDependency support multiple dependencies at once

This changes DepGraph::AddDependency into DepGraph::AddDependencies, which takes
in a single child, but a set of parent transactions, making them all dependencies
at once.

This is important for performance. N transactions can have O(N^2) parents combined,
so constructing a full DepGraph using just AddDependency (which is O(N) on its own)
could take O(N^3) time, while doing the same with AddDependencies (also O(N) on its
own) only takes O(N^2).

Notably, this matters for DepGraphFormatter::Unser, which goes from O(N^3) to O(N^2).

Co-Authored-By: Greg Sanders <gsanders87@gmail.com>

sipa committed:
clusterlin: abstract out DepGraph::GetReduced{Parents,Children}

A fuzz test already relies on these operations, and a future commit will need
the same logic too. Therefore, abstract them out into proper member functions,
with proper testing.

sipa committed:
clusterlin: simplify DepGraphFormatter::Ser

This does not change the serialization format.

It turns out that it is unnecessary to keep track of the order of transactions
in the so-far reconstructed DepGraph to decide how far from the end to insert
a new transaction.

sipa committed:
clusterlin: rework DepGraphFormatter::Unser

This commit does not change the serialization format. Its purpose is making a
few changes already in order to reduce the diff size of the later commit that
introduces support for holes in DepGraph.

The previous approach was to immediately construct a transaction as soon as its
feerate was known in a preliminary position, and then undo that, and place it
in the correct position once the position information is known (such that a
deserialization error in between would not result in an inconsistent state).

The new approach is to delay the actual transaction creation until all its
information is known, avoiding the need to undo and redo. This requires a
different means of determining whether dependencies are redundant, but that has
the advantage that a later commit can apply all dependencies at once, reducing
the complexity of deserialization.

sipa committed:
clusterlin: add DepGraph::RemoveTransactions and support for holes in DepGraph

This commits introduces support in DepGraph for the transaction positions to be
non-continuous. Specifically, it adds:
* DepGraph::RemoveTransactions which removes 0 or more positions from a DepGraph.
* DepGraph::Positions() to get a set of which positions are in use.
* DepGraph::PositionRange() to get the highest used position in a DepGraph + 1.

In addition, it extends the DepGraphFormatter format to support holes in a
compatible way (it serializes non-holey DepGraphs identically to the old code,
and deserializes them the same way)

instagibbs commented:
I'm going to hold off reviewing this until I see the TxGraph PR to better understand its intended usage. :+1:

instagibbs reviewed (commented):
First run through.

suggestion for clusterlin_add_dependencies harness: https://github.com/instagibbs/bitcoin/commit/c5c44da5a0a630a6725ca85e90e4c67b2eb1f8bc

I think this gets similar or better coverage and is easier for me to understand? (alternatively, take an extra bit and use that as decision bit to batch or individually add, and have the other depgraph do the opposite?)

edit: ah, I'm hodling it wrong! Deleted question.

instagibbs commented on src/cluster_linearize.h:
```diff
@@ -184,6 +157,34 @@ class DepGraph
+     * Complexity: O(N) where N=TxCount().
+     */
+    void AddDependencies(const SetType& parents, ClusterIndex child) noexcept
+    {
```
Was reconstructing graphs incorrectly leading to inconsistent sets. Better to fail here early.

```Suggestion
    {
         // We can't add relationships to holes
        Assume(parents.IsSubsetOf(m_used));
```
a check like this for `AddDependency` would also be good:

`Assume(SetType::Singleton(parent).IsSubsetOf(m_used));`

instagibbs reviewed (commented):
it's likely the wrong place for the coverage, but I added some coverage for iterative additional and removal of transactions. Take what you think is useful if any:

```
diff --git a/src/test/fuzz/cluster_linearize.cpp b/src/test/fuzz/cluster_linearize.cpp
index be06898a6e..62b5ec2343 100644
--- a/src/test/fuzz/cluster_linearize.cpp
+++ b/src/test/fuzz/cluster_linearize.cpp
@@ -1,19 +1,20 @@
 // Copyright (c) The Bitcoin Core developers
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
 #include <cluster_linearize.h>
 #include <serialize.h>
 #include <streams.h>
 #include <test/fuzz/fuzz.h>
 #include <test/fuzz/FuzzedDataProvider.h>
+#include <test/fuzz/util.h>
 #include <test/util/cluster_linearize.h>
 #include <util/bitset.h>
 #include <util/feefrac.h>
 
 #include <algorithm>
 #include <stdint.h>
 #include <vector>
 #include <utility>
 
 using namespace cluster_linearize;
@@ -308,30 +309,89 @@ FUZZ_TARGET(clusterlin_cluster_serialization)
             if (i == j) continue;
             if (provider.ConsumeBool()) cluster[i].second.Set(j);
         }
     }
 
     // Construct dependency graph, and verify it matches the cluster (which includes a round-trip
     // check for the serialization).
     DepGraph depgraph(cluster);
     VerifyDepGraphFromCluster(cluster, depgraph);
 
+    // Copy to store deleted information
+    DepGraph depgraph_copy(depgraph);
+
     // Remove an arbitrary subset (in order to construct a graph with holes) and verify that it
     // still sanity checks (incl. round-tripping serialization).
     uint64_t del = provider.ConsumeIntegralInRange<uint64_t>(1, (uint64_t{1} << num_tx) - 1);
     TestBitSet setdel;
     for (ClusterIndex i : depgraph.Positions()) {
         if (del & 1) setdel.Set(i);
         del >>= 1;
     }
+    assert(del == 0);
+
     depgraph.RemoveTransactions(setdel);
     SanityCheck(depgraph);
+
+    FastRandomContext rng{ConsumeUInt256(provider)};
+
+    // Arbitary addition or removal order for transactions we're deleting
+    std::vector<ClusterIndex> del_indexes;
+    for (const auto del_index : setdel) {
+        del_indexes.push_back(del_index);
+    }
+
+    // Tracks deletion index vs re-addition index for dependency re-addition
+    std::map<ClusterIndex, ClusterIndex> map_del_add_index;
+
+    // Re-add everything back in; should not fail sanity checks
+    // But this will have transactions "out of order" and missing
+    // connections from children to previously deleted ancestors.
+    std::shuffle(del_indexes.begin(), del_indexes.end(), rng);
+    for (ClusterIndex del_index : del_indexes) {
+        const auto add_index{depgraph.AddTransaction(depgraph_copy.FeeRate(del_index))};
+        map_del_add_index[del_index] = add_index;
+    }
+
+    std::shuffle(del_indexes.begin(), del_indexes.end(), rng);
+    for (ClusterIndex del_index : del_indexes) {
+       depgraph.AddDependencies(depgraph_copy.Ancestors(del_index), map_del_add_index[del_index]);
+    }
+
+    SanityCheck(depgraph);
+
+    // But if we repeat this N times, we should get identical depgraphs
+    for (int i = 0; i < provider.ConsumeIntegralInRange<int>(1, 5); ++i) {
+        depgraph_copy = depgraph;
+        // Batch removal or shuffled iterated removal
+        if (provider.ConsumeBool()) {
+            depgraph.RemoveTransactions(setdel);
+        } else {
+            std::shuffle(del_indexes.begin(), del_indexes.end(), rng);
+            for (const auto del_index : del_indexes) {
+                depgraph.RemoveTransactions(TestBitSet::Singleton(del_index));
+            }
+        }
+        SanityCheck(depgraph);
+
+        std::shuffle(del_indexes.begin(), del_indexes.end(), rng);
+        map_del_add_index.clear();
+        for (ClusterIndex del_index : setdel) {
+            const auto add_index{depgraph.AddTransaction(depgraph_copy.FeeRate(del_index))};
+            map_del_add_index[del_index] = add_index;
+        }
+        std::shuffle(del_indexes.begin(), del_indexes.end(), rng);
+        for (ClusterIndex del_index : setdel) {
+            depgraph.AddDependencies(depgraph_copy.Ancestors(del_index), map_del_add_index[del_index]);
+        }
+        assert(depgraph == depgraph_copy);
+    }
 }
 
 FUZZ_TARGET(clusterlin_depgraph_serialization)
 {
     // Verify that any deserialized depgraph is acyclic and roundtrips to an identical depgraph.
 
     // Construct a graph by deserializing.
     SpanReader reader(buffer);
     DepGraph<TestBitSet> depgraph;
     try {
```

glozow commented on src/cluster_linearize.h:
```diff
@@ -184,6 +212,74 @@ class DepGraph
+     *
+     * Complexity: O(N) where N=Ancestors(i).Count() (which is bounded by TxCount()).
+     */
+    SetType GetReducedParents(ClusterIndex i) const noexcept
```
Is the hope to be able to reconstruct direct parents of a transaction from the txgraph? Even if a direct parent is also an indirect ancestor?

glozow replied:
thought so... how is it intended to be used?

glozow commented:
Sorry, I don't know how I missed the calls within this PR :facepalm: I think I was grepping only on the first commit.

instagibbs commented on src/cluster_linearize.h:
```diff
@@ -124,19 +149,49 @@ class DepGraph
+    /** Remove the specified positions from this DepGraph.
+     *
+     * The specified positions will no longer be part of Positions(), and dependencies with them are
+     * removed. Note that due to DepGraph only tracking ancestors/descendants (and not direct
```
ah, right! I hadn't considered the mining side. Maybe leave something to that effect?

sipa reviewed (commented):
I addressed most of the feedback received. Thanks @instagibbs and @glozow!

In addition, the `DepGraphFormatter::Unser` logic was changed significantly. The result is simpler to follow I think (doesn't involve adding a transaction once and then undoing the insertion and redoing it once the position is known), and reduces the complexity from $\mathcal{O}(n^3)$ to $\mathcal{O}(n^2)$ in the process. For this, the commits were reordered, and their commit messages expanded to explain the rationale.

@instagibbs I have added your testing of re-adding in `clusterlin_cluster_serialization` as a separate commit with some minor changes (mostly to keep the result efficient).

instagibbs commented on src/cluster_linearize.h:
```diff
@@ -184,6 +157,34 @@ class DepGraph
+     *
+     * Complexity: O(N) where N=TxCount().
+     */
+    void AddDependencies(const SetType& parents, ClusterIndex child) noexcept
```
even better

instagibbs commented on src/cluster_linearize.h:
```diff
@@ -124,19 +149,49 @@ class DepGraph
+     */
+    void RemoveTransactions(const SetType& del) noexcept
+    {
+        m_used -= del;
```
If we're allowing it, let's make sure there's overage for it in the fuzzer

instagibbs commented on src/test/fuzz/cluster_linearize.cpp:
```diff
@@ -848,9 +859,16 @@ FUZZ_TARGET(clusterlin_postlinearize_tree)
+        if (depgraph_gen.Positions()[i]) {
+            depgraph_tree.AddTransaction(depgraph_gen.FeeRate(i));
+        } else {
+            // For holes, add a dummy transaction which is deleted below.
```
ok, seems clear to me now, not sure why I was so confused.

instagibbs commented:
looked at `git range-diff master 9f67d6dcdcf7e9c75a93ad78f8902537599f4f6d 4476915bc76b48eff9b5e67fd6bd7647cc12793f`

LGTM, though I haven't re-validated the serializer, trusting the tests of tests. I want to think more about the fuzz targets to see if there are any other easy pickups in coverage.

instagibbs commented on src/cluster_linearize.h:
```diff
@@ -100,21 +112,34 @@ class DepGraph
-    DepGraph(const DepGraph<SetType>& depgraph, Span<const ClusterIndex> mapping) noexcept : DepGraph(depgraph.TxCount())
+    DepGraph(const DepGraph<SetType>& depgraph, Span<const ClusterIndex> mapping, ClusterIndex pos_range) noexcept : entries(pos_range)
     {
-        Assert(mapping.size() == depgraph.TxCount());
```
is this not still true?

instagibbs commented on src/cluster_linearize.h:
```diff
@@ -100,21 +112,34 @@ class DepGraph
             AddDependencies(parents, mapping[i]);
         }
+        // Verify that the provided pos_range was correct (no unused positions at the end).
+        Assume(pos_range == 0 || m_used[pos_range - 1]);
```
if not erroneous, I think a bit stronger
```Suggestion
        Assume((m_used.None() && pos_range == 0) || m_used.Last() == pos_range - 1);
```

instagibbs commented on src/test/util/cluster_linearize.h:
```diff
@@ -235,27 +253,63 @@ struct DepGraphFormatter
+                for (auto pos : reordering) holes.Reset(pos);
+                for (auto pos : holes) {
+                    if (diff == 0) {
+                        reordering.push_back(pos);
```
Haven't looked close enough to the new format, but pushing another element to `reordering` after hitting `SetType::Size` already seems to make the DepGraph construction later odd? Couldn't this make the mapping argument too large?

instagibbs replied:
Ah, missed the line where it was being set to `diff`.

sipa reviewed (commented):
~~@instagibbs I noticed that your extending of clusterlin_cluster_serialization makes the fuzz test 10x slower (when sanitizers are enabled), and I don't understand why.~~

EDIT: wiped, re-cmaked, recompiled, and the slowdown is gone.

instagibbs reviewed (approved):
ACK aab53ddcd8fcbc3c0be0da9383f8e06abe5badda

reviewed via `git range-diff master 4476915bc76b48eff9b5e67fd6bd7647cc12793f aab53ddcd8fcbc3c0be0da9383f8e06abe5badda`

Fuzz coverage seems good now, with additional documentation/asserts where requested. I didn't dive deep into the new un-serializer logic.

Did you figure out the slowdown? I didn't reproduce locally.

glozow reviewed (commented):
Read through the code and it makes sense to me, but haven't tested much. It would be nice to have a fuzz coverage report? @dergoegge

sipa committed:
clusterlin: merge two DepGraph fuzz tests into simulation test

This combines the clusterlin_add_dependency and clusterlin_cluster_serialization
fuzz tests into a single clusterlin_depgraph_sim fuzz test. This tests starts
from an empty DepGraph and performs a arbitrary number of AddTransaction,
AddDependencies, and RemoveTransactions operations on it, and compares the
resulting state with a naive reimplementation.

sipa committed:
clusterlin: remove Cluster type

instagibbs commented on src/test/fuzz/cluster_linearize.cpp:
```diff
@@ -241,103 +241,149 @@ std::vector<ClusterIndex> ReadLinearization(const DepGraph<BS>& depgraph, SpanRe
 
 } // namespace
 
```
`clusterlin: merge several DepGraph fuzz tests into simulation test`

Seems to only merge 2 fwiw, might just enumerate them?

instagibbs commented on src/test/fuzz/cluster_linearize.cpp:
```diff
@@ -241,103 +241,149 @@ std::vector<ClusterIndex> ReadLinearization(const DepGraph<BS>& depgraph, SpanRe
-    // Verify that any graph of transactions has its ancestry correctly computed by DepGraph, and
-    // if it is a DAG, that it can be serialized as a DepGraph in a way that roundtrips. This
-    // guarantees that any acyclic cluster has a corresponding DepGraph serialization.
+    /** Read any set of transactions from the provider. */
```
```Suggestion
    /** Read any set of transactions from the provider including empty slots. */
```

instagibbs commented on src/test/fuzz/cluster_linearize.cpp:
```diff
@@ -241,103 +241,149 @@ std::vector<ClusterIndex> ReadLinearization(const DepGraph<BS>& depgraph, SpanRe
+     *  sim[i]->first is its individual feerate, and sim[i]->second is its set of ancestors. */
+    std::array<std::optional<std::pair<FeeFrac, TestBitSet>>, TestBitSet::Size()> sim;
+    /** The number of non-nullopt position in sim. */
+    ClusterIndex num_tx{0};
```
mu-nit: `s/num_tx/sim_num_tx/`

instagibbs commented on src/test/fuzz/cluster_linearize.cpp:
```diff
@@ -241,103 +241,149 @@ std::vector<ClusterIndex> ReadLinearization(const DepGraph<BS>& depgraph, SpanRe
+                    break;
+                }
+            }
+            // Update sim.
```
In case somehow the sim had no non-empty slots above and misses the `idx == i` assert
```Suggestion
            // Update sim.
            assert(!sim[i].has_value());
```

instagibbs replied:
d'oh

instagibbs reviewed (approved):
reACK 7051fcdda6fdc95677a34d5c14321d7580eceed8

via `git range-diff master aab53ddcd8fcbc3c0be0da9383f8e06abe5badda 7051fcdda6fdc95677a34d5c14321d7580eceed8`

New fuzzer harness is a nice generalization of what existed prior.

instagibbs commented:
reACK 2b21aebd9754c503bac12d1dbf566b3aa27451e8

via `git range-diff master 7051fcdda6fdc95677a34d5c14321d7580eceed8 2b21aebd9754c503bac12d1dbf566b3aa27451e8`

sdaftuar reviewed (approved):
Code review ACK, just some nits.

sdaftuar commented:
ACK 2b21aebd9754c503bac12d1dbf566b3aa27451e8

ismaelsadeeq reviewed (commented):
Concept ACK

ismaelsadeeq reviewed (commented):
first pass light code review ACK 2b21aebd9754c503bac12d1dbf566b3aa27451e8

Mostly trying to understand what the code is doing!


comments are just nits

ismaelsadeeq commented on src/cluster_linearize.h:
```diff
@@ -184,6 +184,48 @@ class DepGraph
+    /** Compute the (reduced) set of parents of node i in this graph.
+     *
+     * This returns the minimal subset of the parents of i whose ancestors together equal all of
+     * i's ancestors (unless i is part of a cycle of dependencies). Note that DepGraph does not
+     * store the actual set of parents.
```
nit: can go further with the explanation the comments with something like

```suggestion
     * i's ancestors (unless i is part of a cycle of dependencies). Note that DepGraph does not
     * store set of parents; it has to be reduced from ancestor set.
```

for both the parent and child reducing methods.

ismaelsadeeq reviewed (commented):
li

ismaelsadeeq commented on src/cluster_linearize.h:
```diff
@@ -97,24 +109,50 @@ class DepGraph
+     *
+     * @param depgraph   The original DepGraph that is being remapped.
+     *
+     * @param mapping    A vector, such that mapping[i] gives the position in the new DepGraph
```
much better 👍🏾

instagibbs reviewed (approved):
reACK 0b3ec8c59b2b6d598531cd4e688eb276e597c825

via `git range-diff master 2b21aebd9754c503bac12d1dbf566b3aa27451e8 0b3ec8c59b2b6d598531cd4e688eb276e597c825`

ismaelsadeeq reviewed (commented):
reACK 0b3ec8c59b2b6d598531cd4e688eb276e597c825

sdaftuar commented:
reACK 0b3ec8c59b2b6d598531cd4e688eb276e597c825

glozow commented on src/cluster_linearize.h:
```diff
@@ -634,22 +706,26 @@ class SearchCandidateFinder
     {
-        // Determine reordering mapping, by sorting by decreasing feerate.
-        std::iota(m_sorted_to_original.begin(), m_sorted_to_original.end(), ClusterIndex{0});
+        // Determine reordering mapping, by sorting by decreasing feerate. Unusued positions are
```
tiny typo unusued

glozow commented on src/cluster_linearize.h:
```diff
@@ -19,14 +19,6 @@
 
 namespace cluster_linearize {
 
-/** Data type to represent cluster input.
```
Nice, no more naming clash for `Cluster` in my mind

glozow reviewed (commented):
ACK 0b3ec8c59b2b6d598531cd4e688eb276e597c825, reviewed range-diff from  aab53ddcd8fcbc3c0be0da9383f8e06abe5badda and `clusterlin_depgraph_sim`
